contributors/design-proposals/scheduling/scheduling-framework.md

```diff
@@ -0,0 +1,406 @@
+   invoked at other points, for example, they cannot be called before running
+   predicate functions.
+1. Every call to the extenders involves marshaling and unmarshaling JSON.
+   Calling a webhook (HTTP request) is also slower than calling native functions.
+1. It is hard to inform an extender that scheduler has aborted scheduling of
+   a Pod. For example, if an extender provisions a cluster resource and
+   scheduler contacts the extender and asks it to provision an instance of the
+   resource for the Pod being scheduled and then scheduler faces errors
+   scheduling the Pod and decides to abort the scheduling, it will be hard to
+   communicate the error with the extender and ask it to undo the provisioning
+   of the resource.
+1. Since current extenders run as a separate process, they cannot use
+   scheduler's cache. They must either build their own cache from the API
+   server or process only the information they receive from the default scheduler.
+
+The above limitations hinder building high performance and versatile scheduler
+extensions. We would ideally like to have an extension mechanism that is fast
+enough to allow keeping a bare minimum logic in the scheduler core and convert
+many of the existing features of default scheduler, such as predicate and
+priority functions and preemption into plugins. Such plugins will be compiled
+with the scheduler. We would also like to provide an extension mechanism that do
+not need recompilation of scheduler. The expected performance of such plugins is
+lower than in-process plugins. Such out-of-process plugins should be used in
+cases where quick invocation of the plugin is not a constraint.
+
+# OVERVIEW
+
+Scheduler v2 allows both built-in and out-of-process extenders. This new
+architecture is a scheduling framework that exposes several extension points
+during a scheduling cycle. Scheduler plugins can register to run at one or more
+extension points.
+
+#### Non-goals
+
+-  We will keep Kubernetes API backward compatibility, but keeping scheduler
+   v1 backward compatibility is a non-goal. Particularly, scheduling policy
+   config and v1 extenders won't work in this new framework.
+-  Solve all the scheduler v1 limitations, although we would like to ensure
+   that the new framework allows us to address known limitations in the future.
+-  Provide implementation details of plugins and call-back functions, such as
+   all of their arguments and return values.
+
+# DETAILED DESIGN
+
+## Bare bones of scheduling
+
+Pods that are not assigned to any node go to a scheduling queue and sorted by
+order specified by plugins (described [here](#heading=h.dggq0ff44y2y)). The
+scheduling framework picks the head of the queue and starts a **scheduling
+cycle** to schedule the pod. At the end of the cycle scheduler determines
+whether the pod is schedulable or not. If the pod is not schedulable, its status
+is updated and goes back to the scheduling queue. If the pod is schedulable (one
```

Comment:
`goes back to the scheduling queue` or based on priority could it preempt one of the existing pods? Or did I miss something?

Reply:
The scheduling framework does not have preemption logic of its own. Preemption must be built as a plugin installed at "post filter".